proposed/swift-interop.md

```diff
@@ -122,6 +118,12 @@ When calling a function that returns an opaque struct, the Swift ABI always requ
 
 At the lowest level of the calling convention, we do not consider Library Evolution to be a different calling convention than the Swift calling convention. Library Evolution requires that some types are passed by a pointer/reference, but it does not fundamentally change the calling convention. Effectively, Library Evolution forces the least optimizable choice to be taken at every possible point. As a result, we should not handle Library Evolution as a separate calling convention and instead we can manually handle it at the projection layer.
 
+##### Automatic Reference Counting and Lifetime Management
+
+Swift has a strongly-defined lifetime and ownership model. This model is specified in the Swift ABI and is similar to Objective-C's ARC (Automatic Reference Counting) system. When .NET calls into Swift, the .NET GC is responsible for managing all managed objects. Unmanaged objects from C# should either implement `IDisposable` or utilize a designated thin wrapper over the Swift memory allocator, currently accessible through the `NativeMemory` class, to explicitly release memory. It's important to ensure that when a Swift callee function allocates an object that is returned to .NET, the memory is not dereferenced after the call returns. In another scenario, when Swift calls into .NET, a developer in Swift should explicitly dereference all objects after the call to release memory. If a C# managed object is allocated and returned to Swift, the .NET GC will eventually collect it, but Swift will keep track using ARC, which represents an invalid case and should be handled by projection tools. When a C# unmanaged object is returned to Swift, it is treated as a regular Swift object through ARC, and Swift releases the memory by dereferencing it.
```

Comment:
> It's important to ensure that when a Swift callee function allocates an object that is returned to .NET, the memory is not dereferenced after the call returns.

All objects? This would seem to indicate a Swift function can't allocate any global state and return it to .NET and store it. For example an array. I don't know Swift so it is possible that is true. Can an opaque handle not be returned?

> In another scenario

I think this needs some clarity. It is possible the issue is that it is superfluous and is distracting me. Can we just drop it?

Reply:
> All objects? This would seem to indicate a Swift function can't allocate any global state and return it to .NET and store it. For example an array. I don't know Swift so it is possible that is true. Can an opaque handle not be returned?

Not all 👍 . Only "unsafe" or "raw" pointer types, such as UnsafeMutablePointer and UnsafeRawPointer where you need explicit control over memory.



> I think this needs some clarity. It is possible the issue is that it is superfluous and is distracting me. Can we just drop it?

Yes, we can drop it. I wanted to sheds light to `If a C# managed object is allocated and returned to Swift, the .NET GC will eventually collect it, but Swift will keep track using ARC, which represents an invalid case and should be handled by projection tools`.